api/v1beta1/oscmachine_webhook.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var oscmachinelog = logf.Log.WithName("oscmachine-resource")
+
+func (r *OscMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).For(r).Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-infrastructure-cluster-x-k8s-io-v1beta1-oscmachine,mutating=true,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=oscmachines,verbs=create;update,versions=v1beta1,name=moscmachine.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &OscMachine{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (m *OscMachine) Default() {
+	oscmachinelog.Info("default", "name", m.Name)
+
+}
+
+//+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1beta1-oscmachine,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=oscmachines,verbs=create;update,versions=v1beta1,name=voscmachine.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &OscMachine{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *OscMachine) ValidateCreate() error {
+	oscmachinelog.Info("validate create", "name", m.Name)
+	if allErrs := ValidateOscMachineSpec(m.Spec); len(allErrs) > 0 {
+		oscmachinelog.Info("validate error", "error", allErrs)
+		return apierrors.NewInvalid(GroupVersion.WithKind("OscMachine").GroupKind(), m.Name, allErrs)
+	}
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *OscMachine) ValidateUpdate(oldRaw runtime.Object) error {
+	oscmachinelog.Info("validate update", "name", m.Name)
+	var allErrs field.ErrorList
+	old := oldRaw.(*OscMachine)
+
+	oscmachinelog.Info("validate update old imageId", "old imageId", old.Spec.Node.Vm.ImageId)
+	oscmachinelog.Info("validate update imageId", "imageId", m.Spec.Node.Vm.ImageId)
+	if !reflect.DeepEqual(m.Spec.Node.Vm.ImageId, old.Spec.Node.Vm.ImageId) {
```

Comment:
Here, `==` is enough because it is not pointer or complex structure. I do not know if it is bad to intensively use `reflect.DeepEqual`